clang/utils/TableGen/ClangAttrEmitter.cpp

```diff
@@ -898,23 +898,29 @@ namespace {
   }
 
   class EnumArgument : public Argument {
-    std::string type;
+    std::string fullType;
+    StringRef shortType;
     std::vector<StringRef> values, enums, uniques;
+    bool isExternal;
 
   public:
     EnumArgument(const Record &Arg, StringRef Attr)
-        : Argument(Arg, Attr), type(std::string(Arg.getValueAsString("Type"))),
-          values(Arg.getValueAsListOfStrings("Values")),
+        : Argument(Arg, Attr), values(Arg.getValueAsListOfStrings("Values")),
           enums(Arg.getValueAsListOfStrings("Enums")),
-          uniques(uniqueEnumsInOrder(enums)) {
+          uniques(uniqueEnumsInOrder(enums)),
+          isExternal(Arg.getValueAsBit("External")) {
+      StringRef Type = Arg.getValueAsString("Type");
+      shortType = isExternal ? getAttrName().rsplit("::").second : Type;
```

Comment:
An assert that `shortType` contains only identifier characters would be a nice improvement here.  And fullType is only identifier and `:` characters (or even better, identifier and `::` tokens).

Reply:
Do we have a utility for checking if a character is an identifier character somewhere?

Reply:
Looks like `isValidAsciiIdentifier` from `clang/Basic/CharInfo.h` will do the trick

Reply:
Oh, I can't use that here - clang-tblgen can't depend on clang components.

Reply:
Ah, hrmph... not that I know of.  Perhaps this ends up being a bit more annoying than I'd hoped, feel free to do it in a followup, but a 'negative-list' of common typo values (`;`, `:`, `.`, ` `, `$`, etc) might be valuable?

Reply:
Since we do compile the generated code right away it should fail to compile anyway, so I think all this would really do is improve the error someone gets if they make a mistake. I think I'll probably just leave it as is unless you feel strongly that we should do something here.